library/common/http/dispatcher.cc

```diff
@@ -0,0 +1,108 @@
+    : stream_(stream), observer_(observer), http_dispatcher_(http_dispatcher) {}
+
+void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool end_stream) {
+  ENVOY_LOG(debug, "response headers for stream (end_stream={}):\n{}", end_stream, *headers);
+  if (end_stream) {
+    http_dispatcher_.removeStream(stream_);
+  }
+  observer_.on_headers_f(Utility::transformHeaders(*headers), end_stream, observer_.context);
+}
+
+void Dispatcher::DirectStreamCallbacks::onData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "response data for stream (length={} end_stream={})", data.length(), end_stream);
+  if (end_stream) {
+    http_dispatcher_.removeStream(stream_);
+  }
+  observer_.on_data_f(Envoy::Buffer::Utility::transformData(data), end_stream, observer_.context);
+}
+
+void Dispatcher::DirectStreamCallbacks::onTrailers(HeaderMapPtr&& trailers) {
+  ENVOY_LOG(debug, "response trailers for stream:\n{}", *trailers);
+  http_dispatcher_.removeStream(stream_);
+  observer_.on_trailers_f(Utility::transformHeaders(*trailers), observer_.context);
+}
+
+void Dispatcher::DirectStreamCallbacks::onReset() {
+  http_dispatcher_.removeStream(stream_);
+  observer_.on_error_f({ENVOY_STREAM_RESET, {0, nullptr}}, observer_.context);
+}
+
+Dispatcher::DirectStream::DirectStream(AsyncClient::Stream& underlying_stream,
+                                       DirectStreamCallbacksPtr&& callbacks)
+    : underlying_stream_(underlying_stream), callbacks_(std::move(callbacks)) {}
+
+Dispatcher::Dispatcher(Event::Dispatcher& event_dispatcher,
+                       Upstream::ClusterManager& cluster_manager)
+    : current_stream_id_(0), event_dispatcher_(event_dispatcher),
+      cluster_manager_(cluster_manager) {}
+
+envoy_stream_t Dispatcher::startStream(envoy_observer observer) {
+  envoy_stream_t new_stream_id = current_stream_id_++;
+
+  event_dispatcher_.post([this, observer, new_stream_id]() -> void {
+    DirectStreamCallbacksPtr callbacks =
+        std::make_unique<DirectStreamCallbacks>(new_stream_id, observer, *this);
+    AsyncClient& async_client = cluster_manager_.httpAsyncClientForCluster("egress_cluster");
+    AsyncClient::Stream* underlying_stream = async_client.start(*callbacks, {});
+
+    if (!underlying_stream) {
+      callbacks->onReset();
+    } else {
+      DirectStreamPtr direct_stream =
+          std::make_unique<DirectStream>(*underlying_stream, std::move(callbacks));
+      streams_.emplace(new_stream_id, std::move(direct_stream));
+      ENVOY_LOG(debug, "started stream [{}]", new_stream_id);
+    }
+  });
+
+  return new_stream_id;
+}
+
+envoy_status_t Dispatcher::sendHeaders(envoy_stream_t stream_id, envoy_headers headers,
+                                       bool end_stream) {
+  event_dispatcher_.post([this, stream_id, headers, end_stream]() -> void {
+    DirectStream* direct_stream = getStream(stream_id);
+    if (direct_stream != nullptr) {
```

Comment:
When does this happen? Should this be an ASSERT and/or failure?

Reply:
Ping on this question.

Reply:
Nevermind I see your comment above, but presumably this only happens during a cancellation/failure race condition? Can you expand the comment on why/when this would actually happen?

Reply:
Sorry, answered in an inline comment as it seemed valuable to preserve in comment: 

```
    // If direct_stream is not found, it means the stream has already closed or been reset
    // and the appropriate callback has been issued to the caller. There's nothing to do here
    // except silently swallow this.
```

Reply:
Additional to that there is the even simpler case, which is that the caller might have called with a bogus stream_id.

So even when we cover the threading issues and resolve the potential cancellation/failure race condition we still have to decide what to do with bogus input. Do we want to silently swallow, or ASSERT that the pointer should never be null?

Reply:
I think we need to sort out the error handling semantics and be consistent. IMO you should return some type of error in this case. I would add a TODO to track this.

Reply:
@mattklein123 agreed. I added an inline comment and opened a tracking issue.